`exp init`: UX updates (#3430)
iterative/dvc.org: content/docs/command-reference/exp/init.md

## Description

This command helps you get started with DVC Experiments quickly. It reduces
repetitive DVC procedures by creating a necessary `dvc.yaml` file, which assumes
standard locations of your inputs (data, <abbr>parameters</abbr>, and source
code) and outputs (models, <abbr>metrics</abbr>, and
[plots](/doc/command-reference/plots)).

These locations can be customized through the [command options](#options) or via
[configuration](/doc/command-reference/config#exp). Default project structure:

```
├── data/
├── metrics.json
├── models/
├── params.yaml
├── plots/
└── src/
```

The only required argument is the terminal command that runs your experiment(s).
It can be provided directly [as an argument](#the-command-argument) or by using
the `--interactive` (`-i`) mode (which will prompt for it). The command will be
wrapped as a <abbr>stage</abbr> that `dvc exp run` can execute.

<admon type="tip">

A special deep learning `--type` of stage is supported (`dl`), which uses

> [DVCLive](/doc/dvclive) to monitor [checkpoints] during training of ML models.

Reviewer: Or we could get rid of this tip.

## Description

This command helps you get started with DVC Experiments quickly. It reduces
repetitive DVC procedures by creating a necessary `dvc.yaml` file that assumes
standard locations of your inputs (data, <abbr>parameters</abbr>, and source
code) and outputs (models, <abbr>metrics</abbr> and
[plots](/doc/command-reference/plots)).

These locations can be customized through the [command options](#options) or via
[configuration](/doc/command-reference/config#exp). Default project structure
assumed:

```
├── data/
├── metrics.json
├── models/
├── params.yaml
├── plots/
└── src/
```

You must always provide the terminal command that runs your experiment(s). It
can be given either directly [as an argument](#the-command-argument), or by
using the `--interactive` (`-i`) mode, which will prompt you for it. The command
will be wrapped as a <abbr>stage</abbr> that `dvc exp run` can execute.

<admon type="tip">

A special deep learning `--type` of stage is supported (`dl`), which uses
[DVCLive](/doc/dvclive) to monitor [checkpoints] during training of ML models.

</admon>

> `dvc exp init` also generates parts of the boilerplate repo structure: only
> input files and directories by default, or all locations determined in
> interactive mode.

Reviewer: ```suggestion
`dvc exp init` also generates the boilerplate repo structure: input files and directories, 
and parent directories of outputs (the outputs themselves must be generated by the 
experiment command).
```

Reviewer 2: I rephrased this a little.

## Description

`dvc exp init` helps you get started with DVC Experiments quickly. It reduces
repetitive DVC procedures by creating a `dvc.yaml` file that assumes standard
locations of your input data, <abbr>parameters</abbr>, source code, models,
<abbr>metrics</abbr> and [plots](/doc/command-reference/plots) (generating the

> boilerplate structure if needed). These locations can be customized through the

Reviewer: > (generating the
boilerplate structure if needed)

This is only true of the inputs, which is one reason I suggested rewording the intro above to specify "inputs" and "outputs" (and why I think it's hard to separate #3440).

Reviewer 2: > This is only true of the inputs

What about metrics.json, models/ , and plots/ @dberenbaum ? Sure, it's up to the user where to output any other artifact. Should we just note that?

Reviewer 2: Oh I see. Those are only created with in `-i` mode... ⌛

Reviewer 2: OK I changed the Description now. Actually there's even more nuance because with `-i` mode only the locations specified are created. Even input locations can be omitted and thus not generated. PTAL

> `dvc exp init` helps you get started with DVC Experiments quickly. It reduces

Reviewer: Suggested rewording:

> It creates an experiment pipeline in a `dvc.yaml` file, which assumes standard locations for the inputs (source code, data, and parameters) and outputs (model, metrics, and plots) of your model training command. It also generates the input locations if they don't exist.

Reviewer 2: I like this @dberenbaum . I would make some edits though. Let's move it to a separate PR I think... ⌛

> boilerplate DVC procedures by creating a `dvc.yaml` file that assumes standard

Reviewer 2: ```suggestion
repetitive DVC procedures by creating a `dvc.yaml` file that assumes standard
```

locations of your input data, <abbr>parameters</abbr>, source code, models,

> <abbr>metrics</abbr> and [plots](/doc/command-reference/plots). It also

Reviewer: `<abbr>metrics</abbr>` doesn't seem to be working. Should `abbr` be working and supported for both `metrics` and `plots` here @jorgeorpinel @iesahin?

Reviewer 3: It looks like I never started the initial P.R. #2644

Reviewer 2: Thanks for the reminder. It'll work eventually 😬

> <abbr>metrics</abbr> and [plots](/doc/command-reference/plots). It also
> generates boilerplate structure for your input data, parameters and source code
> if they don't already exist. These locations can be customized through the

Reviewer 2: ```suggestion
<abbr>metrics</abbr> and [plots](/doc/command-reference/plots)
(generating the boilerplate structure if needed).
These locations can be customized through the
```

```dvc
$ dvc exp init --interactive
Command to execute: python src/train.py

Enter experiment dependencies.
Path to a code file/directory [src, n to omit]: src/train.py
Path to a data file/directory [data, n to omit]: data/features
Path to a parameters file [params.yaml, n to omit]:

Enter experiment outputs.

> Path to a model file/directory [models, n to omit]: models/predict.h5

Reviewer: This could be in a separate PR, but we should probably use `predict.h5` or `model.h5` instead of `models/predict.h5` until https://github.com/iterative/dvc/issues/5802 is resolved.

Reviewer 2: Not sure. a) it's a bug I think (shouldn't doc based on that) and b) the output of `exp init` does say 

```
Ensure your experiment command creates ... models/predict.h5.
```

Reviewer: Sorry, I should have clarified. This is what happens now:

```
$ dvc exp init -i
Command to execute: python src/train.py

Enter experiment dependencies.
Path to a code file/directory [src, n to omit]: src/train.py
'src/train.py' does not exist, the file will be created.
Path to a data file/directory [data, n to omit]: data/features
'data/features' does not exist, the directory will be created.
Path to a parameters file [params.yaml, n to omit]:
'params.yaml' does not exist, the file will be created.

Enter experiment outputs.
Path to a model file/directory [models, n to omit]: models/predict.h5
Path to a metrics file [metrics.json, n to omit]:
Path to a plots file/directory [plots, n to omit]: n

ERROR: unexpected error - [Errno 2] No such file or directory: '/private/tmp/tmprepo/models/.gitignore'

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

> it's a bug I think

Yeah, we need to prioritize fixing this in DVC.

Reviewer 2: K. Marking this PR with `wait` status. Please lmk when to follow up if possible!

```dvc
$ dvc exp init --interactive
Command to execute: python src/train.py

Enter experiment dependencies.
Path to a code file/directory [src, n to omit]: src/train.py
Path to a data file/directory [data, n to omit]: data/features
Path to a parameters file [params.yaml, n to omit]:

Enter experiment outputs.
Path to a model file/directory [models, n to omit]: models/predict.h5
Path to a metrics file [metrics.json, n to omit]:
Path to a plots file/directory [plots, n to omit]: n

> ...

Reviewer: Now that the output is much shorter, can we include it all instead of truncating it?

Reviewer: `- epochs` should also be removed in the `dvc.yaml` below.

Reviewer 2: Added full output. Removed `params` group from dvc.yaml sample.

```dvc
$ dvc exp init --interactive
Command to execute: python src/train.py

Enter experiment dependencies.
Path to a code file/directory [src, n to omit]: src/train.py
Path to a data file/directory [data, n to omit]: data/features
Path to a parameters file [params.yaml, n to omit]:

Enter experiment outputs.
Path to a model file/directory [models, n to omit]: models/predict.h5
Path to a metrics file [metrics.json, n to omit]:
Path to a plots file/directory [plots, n to omit]: n

> Creating dependencies: src/train.py and params.yaml
> Creating train stage in dvc.yaml

Reviewer: ```suggestion
Creating dependencies: src/train.py and params.yaml
Creating output directories: models
Creating train stage in dvc.yaml
```

  deps:
    - data/features
    - src/train.py
  params:

>     - epochs

Reviewer: We should be keeping `params:` but dropping `epochs`. So it should look like:

```yaml
train:
  cmd: python src/train.py
  deps:
    - data/features
    - src/train.py
  params:
  outs:
    - models/predict.h5
  metrics:
    - metrics.json:
        cache: false
```

Reviewer 2: That's the default now? OK, done.
